balloc/balloc.c

```diff
@@ -628,60 +628,63 @@ struct balloc_group_write_cfg {
 struct balloc_groups_write_cfg {
 	struct balloc_group_write_cfg *bgs_bgc;
 	struct m0_balloc              *bgs_bal;
-	m0_bcount_t                    bgs_current;
 	m0_bcount_t                    bgs_max;
 	int                            bgs_rc;
 	struct m0_mutex                bgs_lock;
 };
 
-static void balloc_group_write_next(struct m0_be_tx_bulk  *tb,
-                                    struct m0_be_op       *op,
-                                    void                  *datum,
-                                    void                 **user)
+static void
+balloc_group_write_credit(struct m0_balloc               *bal,
+                          struct m0_be_tx_bulk           *tb,
+                          struct balloc_groups_write_cfg *bgs,
+                          struct m0_be_tx_credit         *credit)
 {
-	struct balloc_groups_write_cfg *bgs = datum;
-	struct balloc_group_write_cfg  *bgc;
+	m0_be_btree_insert_credit(&bal->cb_db_group_extents, 2,
+		M0_MEMBER_SIZE(struct m0_ext, e_start),
+		M0_MEMBER_SIZE(struct m0_ext, e_end), credit);
+	m0_be_btree_insert_credit(&bal->cb_db_group_desc, 2,
+		M0_MEMBER_SIZE(struct m0_balloc_group_desc, bgd_groupno),
+		sizeof(struct m0_balloc_group_desc), credit);
+}
 
-	m0_be_op_active(op);
-	m0_mutex_lock(&bgs->bgs_lock);
-	if (bgs->bgs_rc == 0 && bgs->bgs_current < bgs->bgs_max) {
-		bgc  = &bgs->bgs_bgc[bgs->bgs_current];
+static void balloc_group_work_put(struct m0_balloc               *bal,
+                                  struct m0_be_tx_bulk           *tb,
+                                  struct balloc_groups_write_cfg *bgs)
+{
+	struct balloc_group_write_cfg *bgc;
+	struct m0_be_tx_credit         credit;
+	m0_bcount_t                    i;
+	bool                           put_successful;
+	int                            rc;
+
+	for (i = 0; i < bgs->bgs_max; ++i) {
+		m0_mutex_lock(&bgs->bgs_lock);
+		rc = bgs->bgs_rc;
+		m0_mutex_unlock(&bgs->bgs_lock);
```

Comment:
In the previous code the mutex lock was taken till operation on `bgc`. Is the current mutex correct ?

Reply:
I see

> m0_mutex_lock(&bgs->bgs_lock);

on line 646 in the original code. `bgc` is not used elsewhere concurrently until it's added as a work item for `m0_be_tx_bulk`, so I don't see a place where this might be an issue. If you do - please let me know.